### PR TITLE
Fix chat streaming bug

### PR DESCRIPTION
**Pull request by taldekar**
*Issue #295*

*Description of changes:*
- Introduced a CompletableFuture for the ChatUiRequestListener to manage asynchronous operations more effectively.
- Implemented a chain of thenApply calls on the CompletableFuture to process streaming messages sequentially.
- This approach ensures that each message in the stream is properly handled, even if the Amazon Q view is closed unexpectedly during the streaming process.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by breedloj — APPROVED** [review]

File: plugin/src/software/aws/toolkits/eclipse/amazonq/chat/ChatCommunicationManager.java [hunk 1]
```diff
@@ -148,8 +149,7 @@ private ChatRequestParams addEditorState(final ChatRequestParams chatRequestPara
         // only include files that are accessible via lsp which have absolute paths
         getOpenFileUri().ifPresent(filePathUri -> {
             chatRequestParams.setTextDocument(new TextDocumentIdentifier(filePathUri));
-            getSelectionRangeCursorState()
-                    .ifPresent(cursorState -> chatRequestParams.setCursorState(Arrays.asList(cursorState)));
+            getSelectionRangeCursorState().ifPresent(cursorState -> chatRequestParams.setCursorState(Arrays.asList(cursorState)));
```
Comment (shruti0085) [inline plugin/src/software/aws/toolkits/eclipse/amazonq/chat/ChatCommunicationManager.java:152]: please revert any formatting changes in this PR. That takes away time from reviewing the actual change

File: plugin/src/software/aws/toolkits/eclipse/amazonq/chat/ChatCommunicationManager.java [hunk 2]
```diff
@@ -233,21 +231,22 @@ private void sendErrorToUi(final String tabId, final Throwable exception) {
...
     /*
      * Sends message to Chat UI to show in webview
      */
     public void sendMessageToChatUI(final ChatUIInboundCommand command) {
-        if (chatUiRequestListener != null) {
-            String message = jsonHandler.serialize(command);
-            chatUiRequestListener.onSendToChatUi(message);
-        }
+        String message = jsonHandler.serialize(command);
+        chatUiRequestListenerFuture.thenApply(listener -> {
```
Comment (shruti0085) [inline plugin/src/software/aws/toolkits/eclipse/amazonq/chat/ChatCommunicationManager.java:246]: this approach seems very odd IMO. Traditionally listeners in an event subscriber pattern are not intended to be completable futures. In this case, the webview class is the listener to an event. Why do we need to return the listener on L248?

Comment (shruti0085) [reply plugin/src/software/aws/toolkits/eclipse/amazonq/chat/ChatCommunicationManager.java:246]: on second thought, I think the fix is simply moving the [removeListener call here ](https://github.com/aws/amazon-q-eclipse/blob/55bba2ab47b22b924a342631d8c1d8d6747e87d9/plugin/src/software/aws/toolkits/eclipse/amazonq/views/AmazonQChatWebview.java#L279)to be within the if (canDisposeState) check .
The reason is when the view is closed the listener is unhooked which means messages sent in the time between it is closed and opened can get dropped leading to the view getting stuck in an in-progress state.
Can you try the above suggestion and validate it works as expected?

File: plugin/src/software/aws/toolkits/eclipse/amazonq/chat/ChatCommunicationManager.java (hunk 1 above)
Comment (taldekar) [reply plugin/src/software/aws/toolkits/eclipse/amazonq/chat/ChatCommunicationManager.java:152]: Okay, will do. Sorry!

File: plugin/src/software/aws/toolkits/eclipse/amazonq/chat/ChatCommunicationManager.java (hunk 2 above)
Comment (breedloj) [reply plugin/src/software/aws/toolkits/eclipse/amazonq/chat/ChatCommunicationManager.java:246]: Let's address this as part of the larger refactor to the views @taldekar.

**Pull request merged**